aiida/engine/processes/workchains/restart.py

```diff
@@ -0,0 +1,332 @@
+
+        try:
+            unwrapped_inputs = self.ctx.inputs
+        except AttributeError:
+            raise AttributeError('no process input dictionary was defined in `self.ctx.inputs`')
+
+        # Set the `CALL` link label
+        unwrapped_inputs['metadata']['call_link_label'] = 'iteration_{:02d}'.format(self.ctx.iteration)
+
+        inputs = self._wrap_bare_dict_inputs(self._process_class.spec().inputs, unwrapped_inputs)
+        node = self.submit(self._process_class, **inputs)
+
+        # Add a new empty list to the `errors_handled` extra. If any errors handled registered through the
+        # `register_error_handler` decorator return an `ErrorHandlerReport`, their name will be appended to that list.
+        errors_handled = self.node.get_extra('errors_handled', [])
+        errors_handled.append([])
+        self.node.set_extra('errors_handled', errors_handled)
+
+        self.report('launching {}<{}> iteration #{}'.format(self.ctx.process_name, node.pk, self.ctx.iteration))
+
+        return ToContext(children=append_(node))
+
+    def inspect_process(self):  # pylint: disable=inconsistent-return-statements,too-many-branches
+        """Analyse the results of the previous process and call the handlers when necessary.
+
+        If the process is excepted or killed, the workchain will abort. Otherwise any attached handlers will be called
+        in order of their specified priority. If the process was failed and no handler returns a report, indicating that
+        the error was handled, it is considered an unhandled process failure and the process is relaunched. If this
+        happens twice in a row, the workchain is aborted. In the case that at least one handler returned a report the
+        following matrix determines the logic that is followed:
+
+            Process  Handler    Handler     Action
+            result   report?    exit code
+            -----------------------------------------
+            Success      yes        == 0     Restart
+            Success      yes        != 0     Abort
+            Failed       yes        == 0     Restart
+            Failed       yes        != 0     Abort
+
+        If no handler returned a report and the process finished successfully, the work chain's work is considered done
+        and it will move on to the next step.
+        """
+        node = self.ctx.children[self.ctx.iteration - 1]
+
+        if node.is_excepted:
+            return self.exit_codes.ERROR_SUB_PROCESS_EXCEPTED  # pylint: disable=no-member
+
+        if node.is_killed:
+            return self.exit_codes.ERROR_SUB_PROCESS_KILLED  # pylint: disable=no-member
+
+        # Sort the handlers with a priority defined, based on their priority in reverse order
+        handlers = [handler for handler in self._handlers if handler.priority]
+        handlers = sorted(handlers, key=lambda x: x.priority, reverse=True)
+
+        last_report = None
+
+        for handler in handlers:
```

Comment:
Just a remark that we need to be very careful to what logic we stick here. Because it will shape the handlers to a very large extent. The difficult decision needs to be made when at least one handler returned a non-zero report and at least one returned a zero status.

The possibilities that I see here are the following:

1. If at least one handler returned a non-zero status - consider that the work chain failed, so it should be stopped.

1. If at least one handler returned a zero status - consider that the error was fixed, so the work chain should be restarted.

1. **(current approach)** Decide whether to stop or to continue the work chain based on the last meaningful handler report. Add `do_break` attribute to the ProcessHandlerReport to make the handler **the last one**.

1. Decide whether to stop or to continue the work chain based on the first meaningful handler report. Add `do_continue` attribute to continue iterating the handlers.

1. Add `stop_workchain` attribute to the ErrorHandlerReport that explicitly says whether to stop or not the work chain if the handler failed.

Reply:
Thinking a bit more on this, I would propose the combination of 2. and 5.:

If at least one handler returned a zero-status - restart the work chain. If at least one handler returned non-zero status and has `stop_workchain` set to `True` - stop the work chain. If all meaningful reports have non-zero exit status - stop the work chain.